### DISCUSSION TITLE
Debugging synchronization status

### DISCUSSION THREAD

**Discussion by matmut7** [vlcn-io/cr-sqlite#264] [Q&A] [answered]
Hi,

Thanks for this awesome project!

I am trying it out by coding a small React / Express app. The Express server is virtually identical to the one in `vite-starter`, it does nothing else.

Sometimes changes in the React client cannot be seen in the serverside DB, even though the Server Sent Events stream seems to be open and working. I'm having a hard time debugging this, I am then wondering if either the client or server library could give some more information on the data replication status, and why it would not be working.

**Comment by tantaman** [comment]
Sure, there should be some logging to the client and server paths. I'm thinking of adding [Winston](https://github.com/winstonjs/winston) to both [client](https://stackoverflow.com/questions/51110058/can-winston-logger-be-used-on-the-front-end-for-logging) and server and having it enable when in dev mode.

Also, do you have a repro of this issue I could look at?

**Comment by matmut7** [comment]
For now I don't, the server DB sometimes gets out of sync and I don't understand when this happens yet. I'll eventually try to look into this and make a minimal reproduction.

**Reply by tantaman** [reply]
Only one tab is ever responsible for sync and it should log any errors it encounters sending or receiving data. Anything show up in your JS logs on the client?

**Reply by matmut7** [reply]
No error client side, all I have is the `next tick!` message which I think is a good sign.
I even see the server receiving packets on `/sync/changes` like:

```
server:dev: MESSAGE {
server:dev:   _tag: 0,
server:dev:   toDbid: Uint8Array(16) [
server:dev:     105, 162,  52,  28,  85,
server:dev:      47,  64, 159, 175,  54,
server:dev:     228, 247, 113, 242, 124,
server:dev:     198
server:dev:   ],
server:dev:   fromDbid: Uint8Array(16) [
server:dev:     91,  76,  39,  26, 42, 203,
server:dev:     74, 187, 167,  84, 71,  22,
server:dev:     41, 199, 116, 176
server:dev:   ],
server:dev:   schemaVersion: -8241443614786113615n,
server:dev:   seqStart: [ 24883n, 0 ],
server:dev:   seqEnd: [ 24885n, 0 ],
server:dev:   changes: [
server:dev:     [ 'expenses', '2', 'name', "'Non'", 1n, 24884n ],
server:dev:     [ 'expenses', '2', 'category_id', '1', 1n, 24884n ],
server:dev:     [ 'expenses', '2', 'author_id', '1', 1n, 24884n ],
server:dev:     [ 'expenses', '2', 'amount', '100.0', 1n, 24884n ],
server:dev:     [
server:dev:       'expenses',
server:dev:       '2',
server:dev:       'created_at',
server:dev:       "'2023-07-07 08:32:22'",
server:dev:       1n,
server:dev:       24884n
server:dev:     ],
server:dev:     [
server:dev:       'participant_expense_coefficients',
server:dev:       '1|2',
server:dev:       'coefficient',
server:dev:       '1.0',
server:dev:       1n,
server:dev:       24885n
server:dev:     ]
server:dev:   ]
server:dev: }
server:dev: RET { _tag: 5 }
```

I also see the client receive the server's response to this, it all seems fine except that I can't see any change on my server side SQLite DB. I am simply opening the file in `server/dbs` with an [app like this](https://sqlitebrowser.org/).

**Reply by tantaman** [reply]
What's the output of '.tables' when you open the server db? And what's the name of the server db? It should be a ~16 byte hex string.

**Reply by matmut7** [reply]
The DB's ID is `69a2341c552f409faf36e4f771f27cc6` which is hard coded for now but I think it comes from the generation code in `vite-starter`, but maybe a version from one or two months.

`.tables` output:
```
__crsql_siteid
categories
categories__crsql_clock
crsql_master
crsql_tracked_peers
expenses
expenses__crsql_clock
participant_category_coefficients
participant_category_coefficients__crsql_clock
participant_expense_coefficients
participant_expense_coefficients__crsql_clock
participants
participants__crsql_clock
```

**Reply by matmut7** [reply]
Changing my DB ID for a new hard coded one seems to make it work again. I'll try to see when it goes out of sync again.

**Reply by tantaman** [reply, marked as answer]
What about

`SELECT * FROM expenses__crsql_clock`, `SELECT * FROM expenses` and `SELECT * FROM crsql_changes` on the server db?

I see your primary keys look to be auto increment. That's probably not going to work as many devices will assign the same key to rows that are actually not intended to be the same.

The primary key should be some sort of UUID (e.g., [nanoid](https://www.npmjs.com/package/nanoid)) or something related to the identity of the row.

**Reply by matmut7** [reply]
I'm so sorry I deleted the previous DB when changing to a new ID :sob: 
But many thanks for the auto increment problem, I'll change it and let you know if I have any more problems!
